### PR TITLE
RED-134847 support for active defrag for RedisJSON.

### PR DESCRIPTION
**Pull request by MeirShpilraien**
The PR adds support for active defrag on RedisJSON. A pre condition for this PR is that the following PR's will be megred:

* [Redis defrad module API extentions](https://github.com/redis/redis/pull/13509)
* [redismodule-rs support for active defrag API](https://github.com/RedisLabsModules/redismodule-rs/pull/387)
* [IJSON support for defrag](https://github.com/RedisJSON/ijson/pull/1)

The PR register defrag function on the json datatype and uses the new capability of ISON to defrag the key.

**Notice**:

* Increamental defrag of the json key is **not** support. In order to support it we need to implement the free_effort callback. This is not trivial and even if it was it would have cause the json object to potentially be freed when the GIL is not hold, which violate the assumption of our shared string implementation (it is not thread safe). We leave it for future improvment.

* If we run on a Redis version that do not support the defrag start callback, we can still partially support defrag. In that case the IJSON object will be defraged but the shared strings dictionatrywill not be reinitialize. This basically means that shared strings will not be defraged.

Tests were added to cover the new functionality.

- [ ] Change deps once all the required PR's get merged

### CODE REVIEW

File: redis_json/src/defrag.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    let mut defrag_stats = DEFRAG_STATS.lock(&defrag_ctx);
+    defrag_stats.keys_defrag += 1;
+
+    let mut defrag_allocator = DefragCtxAllocator {
+        defrag_ctx: &defrag_ctx,
+    };
+    let value = value.cast::<*mut RedisJSON<ijson::IValue>>();
+    let new_val = defrag_allocator.realloc_ptr(*value, Layout::new::<RedisJSON<ijson::IValue>>());
+    if !new_val.is_null() {
+        *value = new_val;
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:77]: Its a pointer so no destructor are called

File: redis_json/src/defrag.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    std::ptr::write(
+        &mut (**value).data as *mut ijson::IValue,
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:80]: `.cast_mut` works on pointers, here we create a pointer out of a reference.

**Review by ephraimfeldblum — APPROVED** [review]

File: tests/pytest/test_defrag.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+def enableDefrag(env):
+    # make defrag as aggressive as possible
+    env.cmd('CONFIG', 'SET', 'hz', '100')
+    env.cmd('CONFIG', 'SET', 'active-defrag-ignore-bytes', '1')
+    env.cmd('CONFIG', 'SET', 'active-defrag-threshold-lower', '0')
+    env.cmd('CONFIG', 'SET', 'active-defrag-cycle-min', '99')
+
+    try:
+        env.cmd('CONFIG', 'SET', 'activedefrag', 'yes')
```
Comment (oranagra) [inline tests/pytest/test_defrag.py:15]: @MeirShpilraien i understand that unlike the redis defrag tests, this one isn't actually creating fragmentation and waiting for it to go down, which is ok.
and i understand that it does test all sorts of various types of json objects, and watches for stats about callback cycles, but how does it make sure that the pointers were really moved? what if redis forgets to call the module api callback?

maybe in addition to these, add one test that create holes and waits for them to get filled?

Comment (MeirShpilraien) [reply tests/pytest/test_defrag.py:15]: Modify to tests to actually cause fragmentation and verify that defrag actually took place.

Comment (oranagra) [reply tests/pytest/test_defrag.py:15]: if i read it correctly, you created a lot of jason fields, and then deleted all of them.
what i think you should do:
1. disable active defrag
2. create a lot of jason fields
3. delete all the even ones (or alternatively when you create them, you can create two keys, and interchangeably add fields to both, and then delete one key)
4. verify you managed to create fragmentation beyond a certain threshold.
5. enable active defrag.
6. wait for fragmentation levels to go down

see memefficiency.tcl in redis.
what i wanna see is that fields get actually re-located (not just that the callback is called)

File: Cargo.toml [hunk 4]
```diff
@@ -7,7 +7,7 @@ members = [
 ]
 
 [workspace.dependencies]
-ijson = { git="https://github.com/RedisJSON/ijson", rev="e0119ac74f6c4ee918718ee122c3948b74ebeba8", default_features=false}
+ijson = { git="https://github.com/RedisJSON/ijson", rev="eede48fad51b4ace5043d3e0714f5a65481a065d", default_features=false}
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply Cargo.toml:10]: This is irrelevant for this PR and should be done on another PR

File: redis_json/src/defrag.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+use redis_module_macros::{defrag_end_function, defrag_start_function};
+
+use crate::redisjson::RedisJSON;
+
+#[derive(Default)]
+pub(crate) struct DefragStats {
+    defrag_started: usize,
+    defrag_ended: usize,
+    keys_defrag: usize,
+}
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:21]: No redismodule-rs its an example. Here we count it for debug purposes and tests only.

File: redis_json/src/defrag.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+impl<'dc> DefragAllocator for DefragCtxAllocator<'dc> {
+    unsafe fn realloc_ptr<T>(&mut self, ptr: *mut T, _layout: Layout) -> *mut T {
+        self.defrag_ctx.defrag_realloc(ptr)
+    }
+
+    /// Allocate memory for defrag
+    unsafe fn alloc(&mut self, layout: Layout) -> *mut u8 {
+        self.defrag_ctx.defrag_alloc(layout)
+    }
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:39]: I replied there

File: redis_json/src/defrag.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            ),
+            (
+                RedisValueKey::String("keys_defrag".to_owned()),
+                RedisValue::Integer(defrag_stats.keys_defrag as i64),
+            ),
+        ]
+        .into_iter()
+        .collect(),
+    ))
+}
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:106]: Yes same answer

File: redis_json/src/defrag.rs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    };
+    let value = value.cast::<*mut RedisJSON<ijson::IValue>>();
+    let new_val = defrag_allocator.realloc_ptr(*value, Layout::new::<RedisJSON<ijson::IValue>>());
+    if !new_val.is_null() {
+        std::ptr::write(value, new_val);
+    }
+    std::ptr::write(
+        &mut (**value).data as *mut ijson::IValue,
+        std::ptr::read(*value).data.defrag(&mut defrag_allocator),
+    );
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:82]: 1. We get the `value.data` and read it (without move)
2. We defrag the data
3. We put the result on `value.data`

The defrag works on value.data not on the pointer, this was defrag and step before.

File: redis_json/src/defrag.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iddm) [reply redis_json/src/defrag.rs:21]: Yes, noticed it too late. Thanks.

File: redis_json/src/defrag.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (iddm) [reply redis_json/src/defrag.rs:106]: Thanks.

File: redis_json/src/defrag.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iddm) [reply redis_json/src/defrag.rs:39]: Thanks.

File: tests/pytest/test_defrag.py [hunk 9]
```diff
@@ -90,3 +90,44 @@ def testDefragObject(env):
...
+            env.assertTrue(False, message='Failed waiting for fragmentation')
+            return
+
+    #enable active defrag
+    env.cmd('CONFIG', 'SET', 'activedefrag', 'yes')
+
+    # wait for fragmentation for go down for up to 30 seconds
+    frag = env.cmd('info', 'memory')['mem_fragmentation_ratio']
+    startTime = time.time()
+    while frag < 1.1:
```
Comment (oranagra) [inline tests/pytest/test_defrag.py:127]: shouldn't it be `>`?
p.s. i think `allocator_frag_ratio` is safer to use.

File: redis_json/src/defrag.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    let mut defrag_stats = DEFRAG_STATS.lock(&defrag_ctx);
+    defrag_stats.keys_defrag += 1;
+
+    let mut defrag_allocator = DefragCtxAllocator {
+        defrag_ctx: &defrag_ctx,
+    };
+    let value = value.cast::<*mut RedisJSON<ijson::IValue>>();
+    let new_val = defrag_allocator.realloc_ptr(*value, Layout::new::<RedisJSON<ijson::IValue>>());
+    if !new_val.is_null() {
```
Comment (iddm) [inline redis_json/src/defrag.rs:76]: If the returned pointer is null, shouldn't we just leave the function?

File: tests/pytest/test_defrag.py (hunk 9 above)
Comment (MeirShpilraien) [reply tests/pytest/test_defrag.py:127]: > shouldn't it be >?

Don't we want the fragmentation ratio to go down?

> p.s. i think allocator_frag_ratio is safer to use.

OK, will change it.

Comment (MeirShpilraien) [reply tests/pytest/test_defrag.py:127]: > shouldn't it be >?

Ohh sorry right, and when I used the `allocator_frag_ratio` it actually failed.

File: redis_json/src/defrag.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (iddm) [reply redis_json/src/defrag.rs:82]: Thanks.

File: redis_json/src/defrag.rs (hunk 10 above)
Comment (MeirShpilraien) [reply redis_json/src/defrag.rs:76]: We should still reallocate the inner, it might be that this pointer does not need to defrag but other does..

Comment (iddm) [reply redis_json/src/defrag.rs:76]: Sure, thanks!

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (iddm) [reply Cargo.toml:10]: Sure.

**Review by iddm — APPROVED** [review]

File: tests/pytest/test_defrag.py (hunk 9 above)
Comment (oranagra) [reply tests/pytest/test_defrag.py]: maybe.. seems odd, but no reason to wonder about it too much

**Review by oranagra — COMMENTED** [review]
test LGTM

File: tests/pytest/test_defrag.py [hunk 11]
```diff
@@ -110,12 +110,12 @@ def testDefragBigJsons(env):
     # wait for fragmentation for up to 30 seconds
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply tests/pytest/test_defrag.py:110]: @oranagra we will not wait for 30 seconds, we will wait for up to 30 seconds to see fragmentation and if we do not see it by than we will fail the test.

[the remark this answers is not in the document]
Comment (oranagra) [reply tests/pytest/test_defrag.py:110]: i know.. still saying that there's no need.
IMHO it's misleading, we're just waiting for cron to update stats..
anyway, i don't really care.

**Review by iddm — APPROVED** [review]

**Pull request merged**